tests/kafkatest/tests/streams/streams_optimized_test.py

```diff
@@ -0,0 +1,150 @@
+        self.kafka = KafkaService(self.test_context, num_nodes=3,
+                                  zk=self.zookeeper, topics=self.topics)
+
+        self.producer = VerifiableProducer(self.test_context,
+                                           1,
+                                           self.kafka,
+                                           self.input_topic,
+                                           throughput=1000,
+                                           acks=1)
+
+    def test_upgrade_optimized_topology(self):
+        self.zookeeper.start()
+        self.kafka.start()
+
+        processor1 = StreamsOptimizedUpgradeTestService(self.test_context, self.kafka)
+        processor2 = StreamsOptimizedUpgradeTestService(self.test_context, self.kafka)
+        processor3 = StreamsOptimizedUpgradeTestService(self.test_context, self.kafka)
+
+        processors = [processor1, processor2, processor3]
+
+        # produce records continually during the test
+        self.producer.start()
+
+        # start all processors unoptimized
+        for processor in processors:
+            self.set_topics(processor)
+            processor.CLEAN_NODE_ENABLED = False
+            self.verify_running_repartition_topic_count(processor, 4)
+
+        self.verify_processing(processors, False)
+
+        self.stop_processors(processors)
+
+        # start again with topology optimized
+        for processor in processors:
+            processor.OPTIMIZED_CONFIG = 'all'
+            self.verify_running_repartition_topic_count(processor, 1)
+
+        self.verify_processing(processors, True)
+
+        self.stop_processors(processors)
+
+        self.producer.stop()
+        self.kafka.stop()
+        self.zookeeper.stop()
+
+    @staticmethod
+    def verify_running_repartition_topic_count(processor, repartition_topic_count):
+        node = processor.node
+        with node.account.monitor_log(processor.STDOUT_FILE) as monitor:
+            processor.start()
+            monitor.wait_until('REBALANCING -> RUNNING with REPARTITION TOPIC COUNT=%s' % repartition_topic_count,
+                               timeout_sec=60,
+                               err_msg="Never saw 'REBALANCING -> RUNNING with REPARTITION TOPIC COUNT=%s' message "
+                                       % repartition_topic_count + str(processor.node.account))
+
+    @staticmethod
+    def verify_stopped(processor):
+        node = processor.node
+        with node.account.monitor_log(processor.STDOUT_FILE) as monitor:
+            processor.stop()
+            monitor.wait_until('OPTIMIZE_TEST Streams Stopped',
+                               timeout_sec=60,
+                               err_msg="'OPTIMIZE_TEST Streams Stopped' message" + str(processor.node.account))
+
+    def verify_processing(self, processors, verify_individual_operations):
+        for processor in processors:
+            if not self.all_source_subtopology_tasks(processor):
```

Comment:
In a tiny percentage of test runs one streams instance ends up with all input source tasks, i.e. (`0_0`, `0_1`, `0_2`, `0_3`), so none of the expected operations are processed on that node.  So we check the task assignment and if it's all input source tasks we skip checking this node.

I added this check after noticing some test flakiness.

Reply:
Interesting! I didn't realize we attempt to balance subtopologies over the instances... Is this important for some reason?